tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -126,6 +127,320 @@ def type_signature(self):
+    self._check_arg_is_anonymous_tuple(arg)
+    py_typecheck.check_len(arg.internal_representation, 2)
+    fn_type = arg.type_signature[0]
+    py_typecheck.check_type(fn_type, computation_types.FunctionType)
+    val_type = arg.type_signature[1]
+    py_typecheck.check_type(val_type, computation_types.FederatedType)
+    if all_equal is None:
+      all_equal = val_type.all_equal
+    elif all_equal and not val_type.all_equal:
+      raise ValueError(
+          'Cannot map a non-all_equal argument into an all_equal result.')
+    fn = arg.internal_representation[0]
+    py_typecheck.check_type(fn, pb.Computation)
+    val = arg.internal_representation[1]
+    py_typecheck.check_type(val, list)
+    for v in val:
+      py_typecheck.check_type(v, executor_value_base.ExecutorValue)
+    children = self.executor._target_executors[val_type.placement]
+    fns = await asyncio.gather(*[c.create_value(fn, fn_type) for c in children])
+    results = await asyncio.gather(*[
+        c.create_call(f, v) for c, (f, v) in zip(children, list(zip(fns, val)))
+    ])
+    return FederatingExecutorValue(
+        results,
+        computation_types.FederatedType(
+            fn_type.result, val_type.placement, all_equal=all_equal))
+
+  @tracing.trace
+  async def _zip(self, arg, placement, all_equal):
+    self._check_arg_is_anonymous_tuple(arg)
+    py_typecheck.check_type(placement, placement_literals.PlacementLiteral)
+    cardinality = len(self.executor._target_executors[placement])
+    elements = anonymous_tuple.to_elements(arg.internal_representation)
+    for _, v in elements:
+      py_typecheck.check_type(v, list)
+      if len(v) != cardinality:
+        raise RuntimeError('Expected {} items, found {}.'.format(
+            cardinality, len(v)))
+    new_vals = []
+    for idx in range(cardinality):
+      new_vals.append(
+          anonymous_tuple.AnonymousTuple([(k, v[idx]) for k, v in elements]))
+    children = self.executor._target_executors[placement]
+    new_vals = await asyncio.gather(
+        *[c.create_tuple(x) for c, x in zip(children, new_vals)])
+    return FederatingExecutorValue(
+        new_vals,
+        computation_types.FederatedType(
+            computation_types.NamedTupleType((
+                (k, v.member) if k else v.member
+                for k, v in anonymous_tuple.iter_elements(arg.type_signature))),
+            placement,
+            all_equal=all_equal))
+
+  async def federated_value_at_server(self, arg):
+    return await self._place(arg, placement_literals.SERVER)
+
+  async def federated_value_at_clients(self, arg):
+    return await self._place(arg, placement_literals.CLIENTS)
+
+  async def federated_eval_at_server(self, arg):
+    return await self._eval(arg, placement_literals.SERVER, True)
+
+  async def federated_eval_at_clients(self, arg):
+    return await self._eval(arg, placement_literals.CLIENTS, False)
+
+  async def federated_apply(self, arg):
+    return await self._map(arg)
```

Comment:
smells a bit like these helpers could move to the executor, or at least to eg a base strategy

Reply:
I disagree :) why do you think they should be moved?

Reply:
discussed offline, agreed to move to the base class `IntrinsicStrategy`